### ISSUE TITLE
Application cannot start in Firefox

### ISSUE THREAD

**Issue by LeXXik** [playcanvas/engine#5856]
Forum: https://forum.playcanvas.com/t/back-screen-in-latest-firefox-version-120-0-playcanvas-version-166-3/33964/1

Repro: https://playcanvas.com/project/1167343/overview/roll-a-ball

Launch a game in latest Firefox. A black screen is shown.

Console:
![image](https://github.com/playcanvas/engine/assets/5677782/75d60357-c5c5-4b4c-ae0b-4bfb47c66179)

**Comment by marklundin** [comment]
Is this FF120? I can't replicate

**Comment by LeXXik** [comment]
![image](https://github.com/playcanvas/engine/assets/5677782/3ec20307-4ab8-4bd1-a166-ae5e7cb9111c)

**Comment by kungfooman** [comment]
Interesting, I just suggested PlayCanvas to someone yesterday and he mentioned the same issue, but it worked with WebGL1 for him :thinking:

**Comment by mvaligursky** [comment]
I'm on Mac and it seems to work for me. Is this a Window issue perhaps?

<img width="727" alt="Screenshot 2023-11-27 at 10 43 00" src="https://github.com/playcanvas/engine/assets/59932779/84296ee6-5d34-4dd7-b686-295d6d8ee4cc">

**Comment by LeXXik** [comment]
I've tried to run under Ubuntu (WSL in Windows) and it runs fine.

**Comment by marklundin** [comment]
OK I can replicate this on Windows FF120. Works in engine 165.5, broken in 1.66.2

**Comment by LeXXik** [comment]
Engine examples work fine, except these (black screen as well):
positional
mini stats
obj
video texture
transform feedback
point cloud
point cloud simulation
particles snow / random sprites / anim index
paint mesh
orbit
fly
first person

These take very long time to load:
shader compile
lights physical units

**Comment by mvaligursky** [comment]
I can repro on Windows as well, Firefox 120. I captured this in Spector, and under the hood rendering seems ok to a multi-sampled buffer. At the very end, we resolve this multi-sampled buffer to a single sampled default backbuffer (NULL), maybe that does not work for some reason in FF on Windows. API calls to do it seem correct to me.

![Screenshot 2023-11-27 at 11 01 55](https://github.com/playcanvas/engine/assets/59932779/c927c0e0-cfa2-458b-bd13-6cf287b8dd55)

I suspect that if the multi-sampling was disabled, this would work.

**Comment by mvaligursky** [comment]
> These take very long time to load:
> shader compile
> lights physical units

Yes, the first one is compiling many large shaders, and is expected to take longer on Windows (where DX shader compiler is very slow). This is a banchmark for it really.

The second one has many complex shaders, and so suffers from the same.

Known issue without a solution at the moment. Shared by all WebGL implementations.

**Comment by LeXXik** [comment]
The project runs fine, if Anti-Alias is disabled.

**Comment by mvaligursky** [comment]
> Engine examples work fine, except these (black screen as well):
> positional
> mini stats
> obj
> video texture
> transform feedback
> point cloud
> point cloud simulation
> particles snow / random sprites / anim index
> paint mesh
> orbit
> fly
> first person

Interesting that I think all these examples that do not work use `new pc.Application` and not `new pc.AppBase` that's used by the working examples. Launcher uses `new pc.AppBase` though I believe.

**Comment by mvaligursky** [comment]
Latest Firefox beta in Window: `121.0b3 (64-bit)` has the same issue.
Also, when I enabled transparent canvas, all works.
So it seems to me that Firefox seems to have a problem resolving RGB8 multi-sampled surface to a default framebuffer.
Rendering to RGB8 multisamplewd texture (RenderToTexture examples does it) seems to work fine.

**Comment by mvaligursky** [comment]
created a Firefox bug report: https://bugzilla.mozilla.org/show_bug.cgi?id=1866762

**Comment by mvaligursky** [comment]
The workaround has been released as part of https://github.com/playcanvas/engine/releases/tag/v1.67.1

I'll leave the issue opened to keep an eye on possible Firefox fix, allowing us to enable AA on newer versions.

**Comment by leonidaspir** [comment]
It seems the same issue happens on latest Firefox for Android, user agent:

Mozilla/5.0 (Android 15; Mobile; rv:132.0) Gecko/132.0 Firefox/132.0

Disabling anti-alias fixes the issue.

As a side note we are getting from time to time reports from players on ChromeOS of encountering the exact same issue (black screen, game seems to play on the background). I haven't been able yet to replicate so I can't confirm if it's the same issue there too.

**Comment by mvaligursky** [comment]
adding workaround here: https://github.com/playcanvas/engine/pull/7096

**Comment by mvaligursky** [comment]
By the way, it based on the ticket it seems Firefox will not be fixing this.
Longer term we should be looking at not using RGB8 framebuffer when transparency is disabled, but use RGBA8 format and make sure transparency ends up being 1.0 (maybe by clearing it to 1 and then blocking alpha channel writes or similar). This is likely not very simple though, and so for now we just disable MSAA on affected Firefox versions.

**Comment by mvaligursky** [comment]
This needs to be investigated again .. it might be back or something, even with the workaround
https://forum.playcanvas.com/t/firefox-black-screen-on-mobile-tablet/38426

**Comment by willeastcott** [comment]
Reminds me of the messed up eye projections on AVP unless anitalias is disabled.

**Comment by mvaligursky** [comment]
I tried to reproduce the issue from the forum post [here](https://github.com/playcanvas/engine/issues/5856#issuecomment-2523407025) and no success yet. I tried the example from the post as well as my simple project with AA enabled on these devices, and all works as expected:
- Windows 10, Firefox 132 & 133
- Windows 11, Firefox 132 & 133
- MacOS 15.1.1, Firefox 133
- Android 15, Google Pixel 8 Pro, Firefox 133.0.2

**Comment by mvaligursky** [comment]
Firefox team closed the ticket we created with `wontfix`.
So it's up to us to add a workaround, or keep MSAA disabled on Firefox.
